File: src/lib/domains/deployments/services/deploymentService.ts
/**
 * Deployment Service - Frontend business logic for deployment management
 */

import { invoke } from '@tauri-apps/api/core';
import { logger } from '$lib/domains/shared';
import type { 
	Deployment, 
	DockerContainer,
	DeploymentCreateRequest,
	DeploymentUpdateRequest
} from '../types';
import { DeploymentStatus } from '../types';

export class DeploymentService {
	private static instance: DeploymentService;

	static getInstance(): DeploymentService {
		if (!DeploymentService.instance) {
			DeploymentService.instance = new DeploymentService();
		}
		return DeploymentService.instance;
	}

	/**
	 * Create a new deployment
	 */
	async createDeployment(request: DeploymentCreateRequest): Promise<Deployment> {
		try {
			logger.info('Creating deployment', { context: 'DeploymentService',  
				name: request.name, 
				projectPath: request.projectPath,
				type: request.type
			});
			
			// Convert frontend request to backend format
			const backendRequest = {
				project_id: request.metadata?.projectId || '',
				name: request.name,
				deployment_type: request.type === 'docker' ? 'Docker' : 'Cli',
				sdk_version: request.metadata?.sdkVersion || 'latest', // FUTURE: Add sdkVersion to request type
				project_type: request.projectType,
				project_path: request.projectPath,
				environment: request.environment.variables || {},
				exposed_port: request.exposedPort,
				docker_image_name: request.dockerImageName,
				dockerfile_path: request.dockerfilePath,
				command: request.command,
				working_directory: request.workingDirectory,
			};
			
			const backendDeployment = await invoke<any>('create_deployment_command', {
				request: backendRequest
			});
			
			const deployment = this.convertBackendDeployment(backendDeployment);
			
			logger.info('Deployment created successfully', { context: 'DeploymentService',  
				deploymentId: deployment.id 
			});
			
			return deployment;
		} catch (error) {
			logger.error('Failed to create deployment', {
				context: 'DeploymentService',
				error,
				data: { name: request.name, projectPath: request.projectPath }
			});
			throw error;
		}
	}

	/**
	 * Convert backend deployment to frontend format
	 */
	private convertBackendDeployment(backendDeployment: any): Deployment {
		return {
			...backendDeployment,
			type: backendDeployment.deployment_type === 'Docker' ? 'docker' : 'cli',
			createdAt: new Date(backendDeployment.created_at),
			updatedAt: new Date(backendDeployment.updated_at),
			startedAt: backendDeployment.started_at ? new Date(backendDeployment.started_at) : undefined,
			stoppedAt: backendDeployment.stopped_at ? new Date(backendDeployment.stopped_at) : undefined,
			// Map backend fields to frontend
			dockerImageName: backendDeployment.docker_image_name,
			dockerfilePath: backendDeployment.dockerfile_path,
			workingDirectory: backendDeployment.working_directory,
			processId: backendDeployment.process_id,
			exposedPort: backendDeployment.exposed_port,
		} as Deployment;
	}

	/**
	 * Get all deployments
	 */
	async getDeployments(): Promise<Deployment[]> {
		try {
			logger.info('Getting deployments', { context: 'DeploymentService' });
			
			const backendDeployments = await invoke<any[]>('get_deployments_command');
			const deployments = backendDeployments.map(d => this.convertBackendDeployment(d));
			
			logger.info('Deployments retrieved', { 
				context: 'DeploymentService', 
				data: { count: deployments.length } 
			});
			
			return deployments;
		} catch (error) {
			logger.error('Failed to get deployments', {
				context: 'DeploymentService',
				error
			});
			throw error;
		}
	}

	/**
	 * Get deployment by ID
	 */
	async getDeployment(deploymentId: string): Promise<Deployment> {
		try {
			logger.info('Getting deployment', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			const backendDeployment = await invoke<any>('get_deployment_command', { 
				deploymentId 
			});
			
			if (!backendDeployment) {
				throw new Error('Deployment not found');
			}
			
			const deployment = this.convertBackendDeployment(backendDeployment);
			
			logger.info('Deployment retrieved', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			return deployment;
		} catch (error) {
			logger.error('Failed to get deployment', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Start a deployment
	 */
	async startDeployment(deploymentId: string): Promise<Deployment> {
		try {
			logger.info('Starting deployment', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			const backendDeployment = await invoke<any>('start_deployment_command', { 
				deploymentId 
			});
			
			const deployment = this.convertBackendDeployment(backendDeployment);
			
			logger.info('Deployment started successfully', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			return deployment;
		} catch (error) {
			logger.error('Failed to start deployment', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Stop a deployment
	 */
	async stopDeployment(deploymentId: string): Promise<Deployment> {
		try {
			logger.info('Stopping deployment', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			const backendDeployment = await invoke<any>('stop_deployment_command', { 
				deploymentId 
			});
			
			const deployment = this.convertBackendDeployment(backendDeployment);
			
			logger.info('Deployment stopped successfully', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			return deployment;
		} catch (error) {
			logger.error('Failed to stop deployment', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Delete a deployment
	 */
	async deleteDeployment(deploymentId: string): Promise<void> {
		try {
			logger.info('Deleting deployment', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			await invoke('delete_deployment_command', { deploymentId });
			
			logger.info('Deployment deleted successfully', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
		} catch (error) {
			logger.error('Failed to delete deployment', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Update a deployment
	 */
	async updateDeployment(deploymentId: string, request: DeploymentUpdateRequest): Promise<Deployment> {
		try {
			logger.info('Updating deployment', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			const backendDeployment = await invoke<any>('update_deployment_command', {
				deploymentId,
				request
			});
			
			const deployment = this.convertBackendDeployment(backendDeployment);
			
			logger.info('Deployment updated successfully', { 
				context: 'DeploymentService', 
				data: { deploymentId } 
			});
			
			return deployment;
		} catch (error) {
			logger.error('Failed to update deployment', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Get deployment logs
	 */
	async getDeploymentLogs(deploymentId: string, tail?: number): Promise<string[]> {
		try {
			logger.info('Getting deployment logs', { 
				context: 'DeploymentService', 
				data: { deploymentId, tail } 
			});
			
			const logs = await invoke<string[]>('get_deployment_logs_command', { 
				deploymentId,
				tail
			});
			
			logger.info('Deployment logs retrieved', { 
				context: 'DeploymentService', 
				data: { deploymentId, logCount: logs.length } 
			});
			
			return logs;
		} catch (error) {
			logger.error('Failed to get deployment logs', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Refresh deployment statuses
	 */
	async refreshDeploymentStatuses(): Promise<Deployment[]> {
		try {
			logger.info('Refreshing deployment statuses', { context: 'DeploymentService' });
			
			const backendDeployments = await invoke<any[]>('refresh_deployment_statuses_command');
			const deployments = backendDeployments.map(d => this.convertBackendDeployment(d));
			
			logger.info('Deployment statuses refreshed', { 
				context: 'DeploymentService', 
				data: { count: deployments.length } 
			});
			
			return deployments;
		} catch (error) {
			logger.error('Failed to refresh deployment statuses', {
				context: 'DeploymentService',
				error
			});
			throw error;
		}
	}

	/**
	 * Convert backend container to frontend format
	 */
	private convertBackendContainer(backendContainer: any): DockerContainer {
		return {
			...backendContainer,
			status: backendContainer.status as any,
			ports: Array.isArray(backendContainer.ports) 
				? backendContainer.ports 
				: (backendContainer.ports ? [backendContainer.ports] : []),
			createdAt: (() => {
				if (!backendContainer.created_at || backendContainer.created_at.trim() === '') {
					return new Date();
				}
				const dateStr = backendContainer.created_at.trim();
				// Try parsing as Unix timestamp (seconds) first
				const unixTimestamp = parseInt(dateStr, 10);
				if (!isNaN(unixTimestamp) && unixTimestamp > 0) {
					// If it's a Unix timestamp in seconds, convert to milliseconds
					const date = new Date(unixTimestamp * 1000);
					if (!isNaN(date.getTime())) {
						return date;
					}
				}
				// Try parsing as ISO string or other date format
				const date = new Date(dateStr);
				return isNaN(date.getTime()) ? new Date() : date;
			})(),
			volumes: backendContainer.volumes || [],
			environment: backendContainer.environment || {}
		} as DockerContainer;
	}

	/**
	 * List Docker containers
	 */
	async listContainers(): Promise<DockerContainer[]> {
		try {
			logger.info('Listing containers', { context: 'DeploymentService' });
			
			const backendContainers = await invoke<any[]>('list_containers_command');
			const containers = backendContainers.map(c => this.convertBackendContainer(c));
			
			logger.info('Containers retrieved', { 
				context: 'DeploymentService', 
				data: { count: containers.length } 
			});
			
			return containers;
		} catch (error) {
			logger.error('Failed to list containers', {
				context: 'DeploymentService',
				error
			});
			throw error;
		}
	}

	/**
	 * Get deployment status color
	 */
	getStatusColor(status: DeploymentStatus): string {
		const colors: Record<DeploymentStatus, string> = {
			[DeploymentStatus.CREATING]: 'text-yellow-600',
			[DeploymentStatus.RUNNING]: 'text-green-600',
			[DeploymentStatus.STOPPED]: 'text-gray-600',
			[DeploymentStatus.FAILED]: 'text-red-600',
			[DeploymentStatus.RESTARTING]: 'text-blue-600',
			[DeploymentStatus.REMOVING]: 'text-gray-400',
			[DeploymentStatus.BUILDING]: 'text-yellow-600'
		};
		return colors[status] || 'text-gray-400';
	}

	/**
	 * Get deployment status icon
	 */
	getStatusIcon(status: DeploymentStatus): string {
		const icons: Record<DeploymentStatus, string> = {
			[DeploymentStatus.CREATING]: '🔨',
			[DeploymentStatus.RUNNING]: '🟢',
			[DeploymentStatus.STOPPED]: '⏹️',
			[DeploymentStatus.FAILED]: '❌',
			[DeploymentStatus.RESTARTING]: '🔄',
			[DeploymentStatus.REMOVING]: '🗑️',
			[DeploymentStatus.BUILDING]: '🔨'
		};
		return icons[status] || '❓';
	}

	/**
	 * Format deployment duration
	 */
	formatDuration(createdAt: string, updatedAt: string): string {
		const created = new Date(createdAt);
		const updated = new Date(updatedAt);
		const duration = updated.getTime() - created.getTime();
		
		const seconds = Math.floor(duration / 1000);
		const minutes = Math.floor(seconds / 60);
		const hours = Math.floor(minutes / 60);
		const days = Math.floor(hours / 24);

		if (days > 0) return `${days}d ${hours % 24}h`;
		if (hours > 0) return `${hours}h ${minutes % 60}m`;
		if (minutes > 0) return `${minutes}m ${seconds % 60}s`;
		return `${seconds}s`;
	}

	/**
	 * Validate deployment request
	 */
	validateDeploymentRequest(request: Partial<DeploymentCreateRequest>): string[] {
		const errors: string[] = [];

		if (!request.name?.trim()) {
			errors.push('Name is required');
		}

		if (!request.projectPath?.trim()) {
			errors.push('Project path is required');
		}

		if (!request.projectType) {
			errors.push('Project type is required');
		}

		return errors;
	}

	/**
	 * Generate deployment name from project
	 */
	generateDeploymentName(projectName: string, projectType: string): string {
		const timestamp = new Date().toISOString().slice(0, 19).replace(/:/g, '-');
		return `${projectName}-${projectType}-${timestamp}`;
	}

	/**
	 * Get project type from path
	 */
	detectProjectType(projectPath: string): string {
		// This would be more sophisticated in a real implementation
		// For now, we'll do basic detection based on common files
		if (projectPath.includes('package.json')) return 'node';
		if (projectPath.includes('Cargo.toml')) return 'rust';
		if (projectPath.includes('requirements.txt') || projectPath.includes('pyproject.toml')) return 'python';
		if (projectPath.includes('go.mod')) return 'go';
		return 'unknown';
	}

	/**
	 * Build Docker image with progress tracking
	 */
	async buildDockerImage(
		contextPath: string,
		imageName: string,
		dockerfilePath?: string
	): Promise<string> {
		try {
			logger.info('Building Docker image', { 
				context: 'DeploymentService',
				contextPath,
				imageName,
				dockerfilePath
			});
			
			const result = await invoke<string>('build_docker_image_command', {
				context_path: contextPath,
				image_name: imageName,
				dockerfile_path: dockerfilePath
			});
			
			logger.info('Docker image built successfully', { 
				context: 'DeploymentService',
				imageName,
				result
			});
			
			return result;
		} catch (error) {
			logger.error('Failed to build Docker image', {
				context: 'DeploymentService',
				error,
				data: { contextPath, imageName, dockerfilePath }
			});
			throw error;
		}
	}

	/**
	 * Check if CLI process is running
	 */
	async getProcessStatus(deploymentId: string): Promise<boolean> {
		try {
			logger.info('Checking process status', { 
				context: 'DeploymentService',
				deploymentId
			});
			
			const isRunning = await invoke<boolean>('get_process_status_command', {
				deployment_id: deploymentId
			});
			
			return isRunning;
		} catch (error) {
			logger.error('Failed to check process status', {
				context: 'DeploymentService',
				error,
				data: { deploymentId }
			});
			throw error;
		}
	}

	/**
	 * Start a Docker container
	 */
	async startContainer(containerId: string): Promise<void> {
		try {
			logger.info('Starting container', { 
				context: 'DeploymentService',
				containerId
			});
			
		await invoke('start_container_command', {
			containerId: containerId
		});
			
			logger.info('Container started successfully', { 
				context: 'DeploymentService',
				containerId
			});
		} catch (error) {
			logger.error('Failed to start container', {
				context: 'DeploymentService',
				error,
				data: { containerId }
			});
			throw error;
		}
	}

	/**
	 * Stop a Docker container
	 */
	async stopContainer(containerId: string): Promise<void> {
		try {
			logger.info('Stopping container', { 
				context: 'DeploymentService',
				containerId
			});
			
		await invoke('stop_container_command', {
			containerId: containerId
		});
			
			logger.info('Container stopped successfully', { 
				context: 'DeploymentService',
				containerId
			});
		} catch (error) {
			logger.error('Failed to stop container', {
				context: 'DeploymentService',
				error,
				data: { containerId }
			});
			throw error;
		}
	}

	/**
	 * Remove a Docker container
	 */
	async removeContainer(containerId: string): Promise<void> {
		try {
			logger.info('Removing container', { 
				context: 'DeploymentService',
				containerId
			});
			
		await invoke('remove_container_command', {
			containerId: containerId
		});
			
			logger.info('Container removed successfully', { 
				context: 'DeploymentService',
				containerId
			});
		} catch (error) {
			logger.error('Failed to remove container', {
				context: 'DeploymentService',
				error,
				data: { containerId }
			});
			throw error;
		}
	}
}

export const deploymentService = DeploymentService.getInstance();

File: src-tauri/src/domains/deployments/services/deployment_service.rs
use super::docker_service::{DockerService, Deployment, DeploymentStatus, DeploymentType, EnvironmentConfig};
use super::cli_service::CliService;
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use uuid::Uuid;
use chrono::Utc;

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateDeploymentRequest {
    pub project_id: String,
    pub name: String,
    pub deployment_type: DeploymentType,
    pub sdk_version: String,
    pub project_type: String,
    pub project_path: String,
    pub environment: HashMap<String, String>,
    // Docker-specific fields
    pub exposed_port: Option<u16>,
    pub docker_image_name: Option<String>,
    pub dockerfile_path: Option<String>,
    // CLI-specific fields
    pub command: Option<String>,
    pub working_directory: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateDeploymentRequest {
    pub id: String,
    pub name: Option<String>,
    pub environment: Option<HashMap<String, String>>,
    pub status: Option<DeploymentStatus>,
}

pub struct DeploymentService {
    pub docker_service: DockerService,
    pub cli_service: CliService,
    // In a real application, this would interact with a database
    // For now, we'll use an in-memory store for demonstration
    deployments: std::sync::Mutex<Vec<Deployment>>,
}

impl DeploymentService {
    pub fn new() -> Self {
        Self {
            docker_service: DockerService::new(),
            cli_service: CliService::new(),
            deployments: std::sync::Mutex::new(Vec::new()),
        }
    }

    /// Create a new deployment
    pub async fn create_deployment(&self, request: CreateDeploymentRequest) -> Result<Deployment, String> {
        let deployment_id = Uuid::new_v4().to_string();
        let now = Utc::now();

        match request.deployment_type {
            DeploymentType::Docker => {
                // Check if Docker is available
                if !self.docker_service.is_docker_available().await? {
                    return Err("Docker is not available or not running".to_string());
                }

                // Generate Dockerfile
                let dockerfile_content = self.docker_service.generate_dockerfile(
                    &request.project_type,
                    &request.sdk_version,
                    &request.project_path,
                )?;

                // Use provided dockerfile path or create one
                let dockerfile_path = request.dockerfile_path
                    .unwrap_or_else(|| format!("{}/Dockerfile", request.project_path));
                
                // Write Dockerfile if it doesn't exist
                if !std::path::Path::new(&dockerfile_path).exists() {
                    std::fs::write(&dockerfile_path, dockerfile_content)
                        .map_err(|e| format!("Failed to write Dockerfile: {}", e))?;
                }

                // Build Docker image
                let image_name = request.docker_image_name
                    .unwrap_or_else(|| format!("{}-{}", request.name.to_lowercase().replace(' ', "-"), deployment_id));
                
                self.docker_service.build_image(
                    &request.project_path,
                    &image_name,
                    Some(&dockerfile_path)
                ).await?;

                // Create deployment record
                let deployment = Deployment {
                    id: deployment_id.clone(),
                    project_id: request.project_id,
                    name: request.name,
                    deployment_type: DeploymentType::Docker,
                    status: DeploymentStatus::Building,
                    sdk_version: request.sdk_version,
                    environment: EnvironmentConfig {
                        variables: request.environment,
                    },
                    docker_image_name: Some(image_name),
                    container_id: None,
                    exposed_port: request.exposed_port,
                    dockerfile_path: Some(dockerfile_path),
                    command: None,
                    working_directory: None,
                    process_id: None,
                    logs: Vec::new(),
                    created_at: now.to_rfc3339(),
                    updated_at: now.to_rfc3339(),
                };

                // Store deployment
                self.deployments.lock().unwrap().push(deployment.clone());

                Ok(deployment)
            }
            DeploymentType::Cli => {
                // Validate CLI command
                let command = request.command
                    .ok_or_else(|| "Command is required for CLI deployment".to_string())?;
                
                let working_dir = request.working_directory
                    .unwrap_or_else(|| request.project_path.clone());

                // Create deployment record first (before spawning process)
                let deployment = Deployment {
                    id: deployment_id.clone(),
                    project_id: request.project_id,
                    name: request.name,
                    deployment_type: DeploymentType::Cli,
                    status: DeploymentStatus::Creating,
                    sdk_version: request.sdk_version,
                    environment: EnvironmentConfig {
                        variables: request.environment,
                    },
                    docker_image_name: None,
                    container_id: None,
                    exposed_port: None,
                    dockerfile_path: None,
                    command: Some(command.clone()),
                    working_directory: Some(working_dir.clone()),
                    process_id: None,
                    logs: Vec::new(),
                    created_at: now.to_rfc3339(),
                    updated_at: now.to_rfc3339(),
                };

                // Store deployment
                self.deployments.lock().unwrap().push(deployment.clone());

                Ok(deployment)
            }
        }
    }

    /// Start a deployment
    pub async fn start_deployment(&self, deployment_id: &str) -> Result<Deployment, String> {
        // Get deployment info and release lock
        let deployment_type = {
            let deployments = self.deployments.lock().unwrap();
            let deployment = deployments.iter().find(|d| d.id == deployment_id)
                .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
            deployment.deployment_type.clone()
        }; // Lock is released here

        match deployment_type {
            DeploymentType::Docker => {
                // Get deployment info and release lock
                let (docker_image_name, container_name, ports, volumes, environment_variables) = {
                    let deployments = self.deployments.lock().unwrap();
                    
                    let deployment = deployments.iter().find(|d| d.id == deployment_id)
                        .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
                    
                    let container_name = format!("{}-{}", deployment.name.to_lowercase().replace(' ', "-"), deployment.id);
                    let ports = if let Some(port) = deployment.exposed_port {
                        vec![(port, 3000)]
                    } else {
                        vec![(3000, 3000)]
                    };
                    let volumes = vec![
                        (deployment.project_id.clone(), "/app".to_string())
                    ];
                    let docker_image_name = deployment.docker_image_name.clone()
                        .ok_or_else(|| "Docker image name not found".to_string())?;
                    let environment_variables = deployment.environment.variables.clone();
                    
                    (docker_image_name, container_name, ports, volumes, environment_variables)
                }; // Lock is released here

                let container_id = self.docker_service.run_container(
                    &docker_image_name,
                    &container_name,
                    &ports,
                    &volumes,
                    &environment_variables,
                ).await?;

                // Update deployment with container_id
                let mut deployments = self.deployments.lock().unwrap();
                let deployment = deployments.iter_mut().find(|d| d.id == deployment_id)
                    .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
                    
                deployment.container_id = Some(container_id);
                deployment.status = DeploymentStatus::Running;
                deployment.updated_at = Utc::now().to_rfc3339();

                Ok(deployment.clone())
            }
            DeploymentType::Cli => {
                // Get deployment info and release lock
                let (command, working_dir, environment) = {
                    let deployments = self.deployments.lock().unwrap();
                    let deployment = deployments.iter().find(|d| d.id == deployment_id)
                        .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
                    
                    let command = deployment.command.clone()
                        .ok_or_else(|| "Command not found".to_string())?;
                    let working_dir = deployment.working_directory.clone()
                        .unwrap_or_else(|| deployment.project_id.clone());
                    let environment = deployment.environment.variables.clone();
                    
                    (command, working_dir, environment)
                }; // Lock is released here

                // Spawn CLI process
                let pid = self.cli_service.spawn_process(
                    deployment_id,
                    &command,
                    Some(&working_dir),
                    &environment,
                ).await?;

                // Update deployment with process_id
                let mut deployments = self.deployments.lock().unwrap();
                let deployment = deployments.iter_mut().find(|d| d.id == deployment_id)
                    .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
                    
                deployment.process_id = Some(pid);
                deployment.status = DeploymentStatus::Running;
                deployment.updated_at = Utc::now().to_rfc3339();

                Ok(deployment.clone())
            }
        }
    }

    /// Stop a deployment
    pub async fn stop_deployment(&self, deployment_id: &str) -> Result<Deployment, String> {
        // Get deployment type and relevant info
        let deployment_type = {
            let deployments = self.deployments.lock().unwrap();
            let deployment = deployments.iter().find(|d| d.id == deployment_id)
                .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
            deployment.deployment_type.clone()
        }; // Lock is released here

        match deployment_type {
            DeploymentType::Docker => {
                // Get container_id and release lock
                let container_id = {
                    let deployments = self.deployments.lock().unwrap();
                    deployments.iter()
                        .find(|d| d.id == deployment_id)
                        .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?
                        .container_id.clone()
                }; // Lock is released here
                
                if let Some(container_id) = container_id {
                    self.docker_service.stop_container(&container_id).await?;
                }
            }
            DeploymentType::Cli => {
                // Stop CLI process
                self.cli_service.stop_process(deployment_id).await?;
            }
        }

        // Update deployment status
        let mut deployments = self.deployments.lock().unwrap();
        let deployment = deployments.iter_mut()
            .find(|d| d.id == deployment_id)
            .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
            
        deployment.status = DeploymentStatus::Stopped;
        deployment.updated_at = Utc::now().to_rfc3339();

        Ok(deployment.clone())
    }

    /// Delete a deployment
    pub async fn delete_deployment(&self, deployment_id: &str) -> Result<(), String> {
        // Get deployment type and relevant info
        let deployment_type = {
            let deployments = self.deployments.lock().unwrap();
            let deployment = deployments.iter().find(|d| d.id == deployment_id)
                .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
            deployment.deployment_type.clone()
        }; // Lock is released here

        match deployment_type {
            DeploymentType::Docker => {
                // Get container_id and release lock
                let container_id = {
                    let deployments = self.deployments.lock().unwrap();
                    deployments.iter()
                        .find(|d| d.id == deployment_id)
                        .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?
                        .container_id.clone()
                }; // Lock is released here
                
                // Stop and remove container if running
                if let Some(container_id) = container_id {
                    let _ = self.docker_service.stop_container(&container_id).await;
                    let _ = self.docker_service.remove_container(&container_id).await;
                }
            }
            DeploymentType::Cli => {
                // Stop and cleanup CLI process
                let _ = self.cli_service.stop_process(deployment_id).await;
                self.cli_service.cleanup_process(deployment_id).await;
            }
        }

        // Remove from deployments list
        let mut deployments = self.deployments.lock().unwrap();
        deployments.retain(|d| d.id != deployment_id);
        Ok(())
    }

    /// Get all deployments
    pub fn get_deployments(&self) -> Result<Vec<Deployment>, String> {
        Ok(self.deployments.lock().unwrap().clone())
    }

    /// Get deployment by ID
    pub fn get_deployment(&self, deployment_id: &str) -> Result<Option<Deployment>, String> {
        Ok(self.deployments.lock().unwrap().iter().find(|d| d.id == deployment_id).cloned())
    }

    /// Get deployment logs
    pub async fn get_deployment_logs(&self, deployment_id: &str, tail: Option<usize>) -> Result<Vec<String>, String> {
        // Get deployment type and relevant info
        let deployment_type = {
            let deployments = self.deployments.lock().unwrap();
            let deployment = deployments.iter().find(|d| d.id == deployment_id)
                .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
            deployment.deployment_type.clone()
        }; // Lock is released here

        match deployment_type {
            DeploymentType::Docker => {
                // Get container_id and logs, then release lock
                let container_id = {
                    let deployments = self.deployments.lock().unwrap();
                    let deployment = deployments.iter()
                        .find(|d| d.id == deployment_id)
                        .ok_or_else(|| format!("Deployment with id {} not found", deployment_id))?;
                    deployment.container_id.clone()
                }; // Lock is released here
                
                if let Some(container_id) = container_id {
                    self.docker_service.get_container_logs(&container_id, tail).await
                } else {
                    Ok(Vec::new())
                }
            }
            DeploymentType::Cli => {
                // Get CLI process logs
                Ok(self.cli_service.get_process_logs(deployment_id, tail))
            }
        }
    }

    /// Update deployment
    pub async fn update_deployment(&self, request: UpdateDeploymentRequest) -> Result<Deployment, String> {
        let mut deployments = self.deployments.lock().unwrap();
        
        if let Some(deployment) = deployments.iter_mut().find(|d| d.id == request.id) {
            if let Some(name) = request.name {
                deployment.name = name;
            }
            if let Some(environment) = request.environment {
                deployment.environment.variables = environment;
            }
            if let Some(status) = request.status {
                deployment.status = status;
            }
            deployment.updated_at = Utc::now().to_rfc3339();

            Ok(deployment.clone())
        } else {
            Err(format!("Deployment with id {} not found", request.id))
        }
    }

    /// Get container status for all deployments
    pub async fn refresh_deployment_statuses(&self) -> Result<Vec<Deployment>, String> {
        // Get all deployments with their types and relevant info
        let deployment_info: Vec<(String, DeploymentType, Option<String>, Option<u32>)> = {
            let deployments = self.deployments.lock().unwrap();
            deployments
                .iter()
                .map(|d| (d.id.clone(), d.deployment_type.clone(), d.container_id.clone(), d.process_id))
                .collect()
        }; // Lock is released here
        
        // Check status for each deployment
        let mut status_updates = Vec::new();
        for (deployment_id, deployment_type, container_id, _process_id) in deployment_info {
            match deployment_type {
                DeploymentType::Docker => {
                    if let Some(container_id) = container_id {
                        match self.docker_service.get_container_status(&container_id).await {
                            Ok(status) => {
                                let new_status = match status.as_str() {
                                    "running" => DeploymentStatus::Running,
                                    "exited" => DeploymentStatus::Stopped,
                                    _ => DeploymentStatus::Unknown,
                                };
                                status_updates.push((deployment_id, new_status));
                            }
                            Err(_) => {
                                status_updates.push((deployment_id, DeploymentStatus::Error));
                            }
                        }
                    }
                }
                DeploymentType::Cli => {
                    match self.cli_service.is_process_running(&deployment_id).await {
                        Ok(is_running) => {
                            let new_status = if is_running {
                                DeploymentStatus::Running
                            } else {
                                DeploymentStatus::Stopped
                            };
                            status_updates.push((deployment_id, new_status));
                        }
                        Err(_) => {
                            status_updates.push((deployment_id, DeploymentStatus::Error));
                        }
                    }
                }
            }
        }
        
        // Update deployment statuses
        let mut deployments = self.deployments.lock().unwrap();
        for (deployment_id, new_status) in status_updates {
            if let Some(deployment) = deployments.iter_mut().find(|d| d.id == deployment_id) {
                deployment.status = new_status;
                deployment.updated_at = Utc::now().to_rfc3339();
            }
        }

        Ok(deployments.clone())
    }
}
